File: code/chapter5/logistic_regression.py
import numpy as np
from scipy.optimize import minimize

class LogisticRegression:
    def __init__(self):
        self.coef = None

    # we use '_' as the prefix of the method name to indicate the method is private
    def _sigmoid(self, x):
        return 1.0/(1.0 + np.exp(-x))

    def _log_lik(self, beta, x, y):
        linear = np.dot(x, beta)
        ll = np.sum(linear.dot(y)) - np.sum(np.log(1.0+np.exp(linear)))
        return -ll

    def fit(self, x, y):
        self.coef = minimize(fun=self._log_lik, args=(np.insert(x, 0, 1.0, axis=1),y), method='L-BFGS-B', x0=np.zeros(1+x.shape[1])).x
    
    def predict(self, new_x):
        linear = np.insert(new_x, 0, 1.0, axis=1).dot(self.coef) # add 1 to new_x
        return self._sigmoid(linear)

File: code/chapter5/logistic_regression.R
library(R6)
LogisticRegression = R6Class(
  "LogisticRegression",
  public = list(
    coef = NULL,
    initialize = function() {
    },
    sigmoid = function(x) {
      1.0/(1.0 + exp(-x))
    },
    log_lik = function(beta, x, y) {
      linear = x %*% beta
      ll = sum(linear * y) - sum(log(1 + exp(linear)))
      return(-ll) # return negative log-likelihood
    },
    fit = function(x, y) {
      if (!is.matrix(x)) x = data.matrix(x)
      self$coef = optim(par=rep(0, 1+ncol(x)), fn=self$log_lik, x=cbind(1, x), y=y, method="L-BFGS-B")$par
    },
    predict = function(new_x) {
      if (!is.matrix(new_x)) new_x = data.matrix(new_x)
      linear = cbind(1, new_x) %*% self$coef
      self$sigmoid(linear)
    }
  )
)
